File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int l = 0;
        int r = 1;
        int c = 0;
        int recur = 0;
        while(r<prices.size()){
            int su = prices[r] - prices[l];
            // cout<<su<<" "<<recur<<endl;
            if (prices[l] > prices[r]){
                c += recur;
                l = r;
                r ++ ;
                recur= 0;
            }
            else if (su <= recur){
                c += recur;
                l = r;
                r++;
                recur = 0;
            }
            else{
                recur = su;
                r ++;
            }
        }
        c += recur;
        return c;
    }
};

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        def traverse(indi , buy):
            if indi == len(prices):
                return 0
            if dp[indi][buy] != -1 : return dp[indi][buy]
            if buy == 1:
                profit = max(-prices[indi] + traverse(indi+1 , 0) , traverse(indi+1 , 1))
            else:
                profit = max(prices[indi] + traverse(indi+1,1) , traverse(indi+1 ,0))
            dp[indi][buy] = profit
            return dp[indi][buy]
        dp = [[-1 for i in range(2)] for j in range(len(prices))]

        return traverse(0 , 1)
